core/sequencer/server.go

```diff
@@ -343,21 +341,22 @@ func (s *Server) readMessages(ctx context.Context, source *spb.MapMetadata_Sourc
 	ss := metadata.Source(source)
 	low := ss.StartTime()
 	high := ss.EndTime()
-	// Loop until less than chunkSize items are returned.
-	for count := chunkSize; count == chunkSize; {
-		batch, err := s.logs.ReadLog(ctx, directoryID, source.LogId, low, high, chunkSize)
+	for moreToRead := true; moreToRead; {
+		// Request one more item than chunksize so we can find the next page token
+		batch, err := s.logs.ReadLog(ctx, directoryID, source.LogId, low, high, chunkSize+1)
```

Comment:
It would feel more honest to me if this queried using `chunkSize` without the `+1`. I think the logic still checks out (assuming you don't add the 1 in the check below either).

Reasoning: I assume chunkSize is picked based on what the DB can handle. In the case of the SQL optimizer fun we had recently, I would be somewhat upset if I picked a number under that limit, but then the code between me and the DB started incrementing that size for $reasons, and triggered the issue.

Reply:
I'm not sure chunkSize - 1 would be much better, since it causes issues around the lower end.  

The +1, -1 semantics are bound up in figuring out whether there are more items to read, to distinguish the  difference between an exact batch and there being more items to read.